Hw2

arekalov opened the pull request:
Compose:
1. Весь проект написан на Compose.
2. Создана тема приложения + preview в отдельном файле.
3. Сделаны preview экранов в темной и светлой теме.
4. Анимация elevation toolbar'a.

Coroutines:
1. Все задачи происходят на Dispatcher.Default.
2. Текущие задачи отменяются при убийству viewModel, тк все корутины привязаны к vewModelScope.
3. При возникновении ошибок появляется Toast c message ошибки.
4. Используется Flow.

P. S.  С удалением жестами пока есть проблемы, будет допилено в ближайшее время, тк к данному дз не относится.

arekalov committed:
refactor: edit appBar, change navigation's animations between screens, clean code

Change toolbar: now there is largeToolbar with info about done items,
also add elevation
animation (dpAnimation).
Change navigation
animations duration (it was to long).
Simple refactor in screens files.
Add ThemePreview for demonstration ToDoListTheme.

arekalov committed:
refactor: move all operations with data to background

In repository change simple functions to suspend functions, so viewModel
use launch to call
it on Dispatcher.Default.
Add Toast to show errors on background.

LordGum commented:
Замечаний никаких нет, все сделано красиво и чисто.
могу только добавить, что можно заменить Dispatchers.Default на Dispatchers.IO - так как он покрывает все нужны задачи на добавление и чтение (input & output), а Default для более сложных вычислений в фоне.
большой респект за toolbar 🥇 
compose 4/4
многопоточность 4/4

arekalov commented:
> Замечаний никаких нет, все сделано красиво и чисто. могу только добавить, что можно заменить Dispatchers.Default на Dispatchers.IO - так как он покрывает все нужны задачи на добавление и чтение (input & output), а Default для более сложных вычислений в фоне. большой респект за toolbar 🥇 compose 4/4 многопоточность 4/4

Большое спасибо)

arekalov committed:
refactor: change swipe to dismiss behaviour

Removing work correct, but there is some problems with check
gesture.
Also fix small bugs like aligning visibilityAll icon, add
LineThrough effect for completed task.

NewUserKK commented on app/src/main/java/com/arekalov/yandexshmr/ToDoItemsViewModel.kt:
```diff
@@ -1,52 +1,54 @@
+    private val _items = MutableStateFlow<List<ToDoItem>>(emptyList())
     val items: StateFlow<List<ToDoItem>> = _items.asStateFlow()
+    private val _error = MutableStateFlow<String?>(null)
+    val error: StateFlow<String?> = _error.asStateFlow()
```
nit: будет выглядеть чище если группировать field и backing field вместе, и использовать getter для торчащих наружу:
```
    private val _items = MutableStateFlow<List<ToDoItem>>(emptyList())
    val items: StateFlow<String?>
        get() = _items.asStateFlow()

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?>
        get() = _error.asStateFlow()
```

ждём kotlin 2 с [ExplicitBackingFields](https://noob-programmer.medium.com/kotlins-explicit-backing-fields-a-cleaner-way-to-work-with-mutablestateflow-926e2dbc946e) :)

NewUserKK commented on app/src/main/java/com/arekalov/yandexshmr/ToDoItemsViewModel.kt:
```diff
@@ -1,52 +1,54 @@
+    }
+
+    private fun fetchItems() {
+        viewModelScope.launch(Dispatchers.Default + errorHandler) {
```
1. для целей походов в сеть/чтения из файлов (иначе говоря, задач, которые много ждут ответа от внешних сил) лучше использовать `Dispatchers.IO`. `Dispatchers.Default` предназначен для CPU-сложных задач (пример: перемножение матриц). Но, это не main-поток, так что ок
2. диспетчер можно вынести в поле и избавиться от дублирования, везде его используя

NewUserKK commented on app/src/main/java/com/arekalov/yandexshmr/ToDoItemsViewModel.kt:
```diff
@@ -57,4 +59,14 @@ class ToDoItemsViewModel : ViewModel() {
     fun isItemExists(id: String): Boolean {
         return _items.value.find { it.id == id } != null
```
можно тут хранить MutableMap, а не List чтобы get/has работал за O(1), а не за O(n) -- работать с ней не сильно сложнее, а профит есть

а также, функции по типу exists обычно проще и лучше реализовать через вызов `getItem() != null`

NewUserKK commented on app/src/main/java/com/arekalov/yandexshmr/ToDoItemsViewModel.kt:
```diff
@@ -1,52 +1,54 @@
     private val repository = ToDoItemRepository()
-    private val list = repository.itemList
-    private val _items = MutableStateFlow(list)
+    private val _items = MutableStateFlow<List<ToDoItem>>(emptyList())
```
ещё вариант того, как можно сделать взаимоисключающие стейты (ведь по сути error исключает items, например):
```
sealed interface ToDoItemsState {
    object Loading : ToDoItemsState
    data class Loaded(val items: List<ToDoItem>): ToDoItemsState
    data class Error(val error: Throwable): ToDoItemsState
}
```

и во вью метод
```
fun render(state: ToDoItemsState) = when (state) {
    is Loading -> renderLoad()
    is Loaded -> renderLoaded(state)
    is Error -> renderError(state)
}
```

NewUserKK commented on app/src/main/java/com/arekalov/yandexshmr/models/ToDoItemRepository.kt:
```diff
@@ -1,163 +1,173 @@
+    )
+    private val _todoItems = MutableStateFlow(itemsList)
+    val todoItems: Flow<List<ToDoItem>> = _todoItems.asStateFlow()
+    suspend fun addTodoItem(todo: ToDoItem) = withContext(Dispatchers.Default) {
```
nit: отделить пустой строкой

NewUserKK replied:
> withContext(Dispatchers.Default)

тут есть разные мнения о том, на каком уровне это всё нужно делать

я обычно придерживаюсь подхода, что ViewModel/Facade решает, на каком контексте (потоке) исполнять корутину и это не область ответственности репозитория (если, конечно, реализация внутри не требует явного переключения контекста)

но имеет место быть и перекладывание ответственности за контекст на репозиторий -- тут кому как больше нравится

ещё для более сложного многопоточного кода есть аннотации `@MainThread, @WorkerThread, @AnyThread`, которые подсказывают разработчику, на каком потоке можно безопасно исполнять код -- не призываю тут использовать, просто иметь в виду, что такое существует

NewUserKK commented on app/src/main/java/com/arekalov/yandexshmr/models/ToDoItemRepository.kt:
```diff
@@ -1,163 +1,173 @@
+    suspend fun updateTodoItem(id: String, updatedTodo: ToDoItem) =
+        withContext(Dispatchers.Default) {
+            val currentList = _todoItems.value.toMutableList()
+            val item = currentList.find { it.id == id }
```
same, для удобства можно использовать мапу (и конструировать её как `listOf(...).associateBy { it.id }`)

NewUserKK commented on app/src/main/java/com/arekalov/yandexshmr/models/ToDoItemRepository.kt:
```diff
@@ -1,163 +1,173 @@
+    private val _todoItems = MutableStateFlow(itemsList)
+    val todoItems: Flow<List<ToDoItem>> = _todoItems.asStateFlow()
+    suspend fun addTodoItem(todo: ToDoItem) = withContext(Dispatchers.Default) {
+        val currentList = _todoItems.value.toMutableList()
```
можно проще: `_todoItems.update { it + todo }`

и попробуй для методов ниже сделать по такому же принципу

и вообще, для конструкций `flow.value = doSomething(flow.value)` лучше использовать `flow.update { doSomething(it) }`, потому что это гарантирует, что между двумя вызовами flow.value не произошло изменений в другом потоке

NewUserKK commented on app/src/main/java/com/arekalov/yandexshmr/screens/EditScreen.kt:
```diff
@@ -75,7 +75,7 @@ fun EditScreen(
             } else {
                 isItemNew = true
                 val tempId = LocalDateTime.now().hashCode()
```
можно генерить через java.util.UUID

NewUserKK commented on app/src/main/java/com/arekalov/yandexshmr/screens/EditScreen.kt:
```diff
@@ -208,9 +208,11 @@ fun DeadlinePicker(
-                if (deadline != null) deadline.atStartOfDay()?.toInstant(ZoneOffset.MIN)
-                    ?.toEpochMilli()
-                else null
+                if (deadline != null) {
```
`deadline?.atStartOfDay()?.toInstant(ZoneOffset.MIN)?.toEpochMilli()` ?

NewUserKK commented on app/src/main/java/com/arekalov/yandexshmr/screens/EditScreen.kt:
```diff
@@ -387,7 +395,10 @@ fun PriorityPicker(
                     )
                 },
-                onClick = { onRegularClick(); expanded = false }
+                onClick = {
```
👍

NewUserKK commented on app/src/main/java/com/arekalov/yandexshmr/screens/EditScreen.kt:
```diff
@@ -420,7 +437,12 @@ fun PriorityPicker(
     ToDoListTheme {
-        PriorityPicker(priority = Priority.LOW, {}, {}, {})
+        PriorityPicker(
+            priority = Priority.LOW,
```
👍

NewUserKK commented on app/src/main/java/com/arekalov/yandexshmr/ToDoItemsViewModel.kt:
```diff
@@ -1,52 +1,54 @@
+        fetchItems()
+    }
+
+    private fun fetchItems() {
```
`fetchItems` выглядит как единоразовое действие, а у тебя тут по сути подписка на обновление -- я бы перименовал на что-то вроде startObservingItems()

NewUserKK commented on app/src/main/java/com/arekalov/yandexshmr/screens/HomeScreen.kt:
```diff
@@ -254,12 +321,33 @@ fun HomeScreen(
+            ),
+            Toast.LENGTH_SHORT
+        ).show()
+        toDoItemsViewModel.clearError()
```
тут в случае ошибки у тебя одноразовый тост, в идеале вообще рендерить ошибку на месте списка (а то получается дали тост и отображаем пустой экран, ещё и без возможности перезапроса)

как правило вообще у компонента четыре возможных UI-состояния, которые стоит реализовать: loading, loaded, error, empty

NewUserKK commented on app/src/main/java/com/arekalov/yandexshmr/screens/HomeScreen.kt:
```diff
@@ -294,39 +382,65 @@ fun HomeScreen(
 }
 
-val toDoItemsList = ToDoItemRepository().itemList
+val toDoItemsList = ToDoItemRepository().itemsList
```
на будущее: эти данные для превью вероятно лучше куда-нибудь ещё вынести

NewUserKK reviewed (commented):
Всё круто, всё сделано с доп заданиями, фундаментальных проблем не вижу, все баги из предыдущей домашки пофикшены

Попробуй только нормально выравнять тулбар в collapsed-состоянии -- сейчас заголовок находится выше иконки, и он как-то очень много места занимает

8/8

arekalov committed:
refactor: edit drop to dismiss

Delete check behaviour of endToStart gesture, instead this set delete
behaviour for both gestures.

arekalov committed:
refactor: execution of comments and nits on PR

Edit structures of backing fields in homScreen. Change Dispatcher to
Dispatcher.IO (it was
only IO operations, not hard cpu). Remove withContext in repository (now it's responsibility of viewModel).
